app/domain/authentication/authn_k8s/authenticator.rb

```diff
@@ -1,38 +1,111 @@
+require_relative 'host'
+
 module Authentication
   module AuthnK8s
+
+    # TODO: delete or change these into error classes like the others with self
+    #       contained messages.
+    #
     class AuthenticationError < RuntimeError; end
-    class CSRVerificationError < RuntimeError; end
     class ClientCertVerificationError < RuntimeError; end
     class ClientCertExpiredError < RuntimeError; end
-    class NotFoundError < RuntimeError; end
+
+    AuthenticatorNotFound = ::Util::ErrorClass.new(
+      "'{0}' wasn't in the available authenticators"
+    )
+    WebserviceNotFound = ::Util::ErrorClass.new(
+      "Webservice '{0}' wasn't found"
+    )
+    HostNotFound = ::Util::ErrorClass.new(
+      "Host '{0}' wasn't found"
+    )
+    HostNotAuthorized = ::Util::ErrorClass.new(
+      "'{0}' does not have 'authenticate' privilege on {1}"
+    )
+    CSRIsMissingSpiffeId = ::Util::ErrorClass.new(
+      'CSR must contain SPIFFE ID SAN'
+    )
+    CSRNamespaceMismatch = ::Util::ErrorClass.new(
+      'Namespace in SPIFFE ID must match namespace implied by common name'
+    )
+    PodNotFound = ::Util::ErrorClass.new(
+      "No Pod found for podname '{0}' in namespace '{1}'"
+    )
+    ScopeNotSupported = ::Util::ErrorClass.new(
+      "Resource type '{0}' identity scope is not supported in this version " +
+      "of authn-k8s"
+    )
+    ControllerNotFound = ::Util::ErrorClass.new(
+      "Kubernetes {0} {1} not found in namespace {2}"
+    )
+    CertInstallationError = ::Util::ErrorClass.new(
+      "Cert could not be copied to pod: {0}"
+    )
+    ContainerNotFound = ::Util::ErrorClass.new(
+      "Container {0} was not found for requesting pod"
+    )
     
     class Authenticator
-      def initialize(env:)
-        @env = env
+
+      #TODO: inject Resource, others
+      #
+      def initialize(
```

Comment:
I'm confused about this change. It looks like authenticators only get an env variable by design. Why is this now split out into individual params?